Core/Misc : Creature Scaling System v2.0

Seyden opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

Text copied from Traesh his Pull request :D

**Changes proposed:**

-  Implementing legion creature scaling system.
-  Add new table "creature_template_scaling" with scaling minlevel, maxlevel and deltalevel
-  If the creature has it's entry in "creature_template_scaling", the levels will scale between creature_template_scaling.levelScalingMin and creature_template_scaling.levelScalingMax

**Target branch(es):** master
- [x] master

**Issues addressed:** Closes #17898

Replaces #19264
**Tests performed:** (Does it build, tested in-game, etc.)


**Known issues and TODO list:**


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Traesh committed:
Begin working on scaling system

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Traesh committed:
Work on scaling system

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Conflicts:
	src/server/game/Entities/Creature/Creature.cpp

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Traesh committed:
Set a map of maxhealth by level

Traesh committed:
CREATURE_FLAG_EXTRA_SCALING_LEVEL

Traesh committed:
Handle Logs : SpellNonMeleeDamageLog, SpellPeriodicAuraLog, AttackerStateUpdate

Traesh committed:
Fix compile

Traesh committed:
Handle scaling damages

Traesh committed:
Add targetLevelWithDelta to level for target

Traesh committed:
Handle exp win

Traesh committed:
Fix typo

Traesh committed:
Remove unused var

Traesh committed:
Comment unused param

Traesh committed:
Fix flags_extra in TC_LOG

Traesh committed:
Typo

Traesh committed:
Fix logs for damages creature to player

Traesh committed:
Some fixes & refactoring

Traesh committed:
Reset const

Traesh committed:
Refactor + revert changes on UNIT_FIELD_LEVEL, can now be random between min & max

Traesh committed:
SetLevel(maxlevel)

Traesh committed:
Handle player seeing level -1 if < 50% xp

Traesh committed:
Except for max level

Traesh committed:
Remove flag extra, add creature_template_scaling

Traesh committed:
Remove uint64 flags_extra

Traesh committed:
Fix compile

Traesh committed:
Merge branch 'master' of github.com:Traesh/TrinityCore into scaling

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Traesh committed:
Use SandboxScalingData rather than hardcoding log packet damage

Traesh committed:
Enum instead of magic number + handle TYPE_CREATURE_TO_CREATURE_DAMAGE;

Traesh committed:
Fix travis compile

Traesh committed:
typo

Traesh committed:
Remove wrong TYPE_PLAYER_TO_PLAYER_DAMAGE

Traesh committed:
Switch GenerateDataForUnits to template + remove unused data (infos by soulfrost)

Traesh committed:
TYPE_PLAYER_TO_PLAYER_HEALING => TYPE_PLAYER_TO_PLAYER

Traesh committed:
Fix warning

Traesh committed:
Stupid

Traesh committed:
Fix default value

Traesh committed:
Clarification

Traesh committed:
Merge github.com:TrinityCore/TrinityCore into scaling

Seyden committed:
More creature scaling fixes

Seyden committed:
Merge branch 'master' into scaling

# Conflicts:
#	src/server/game/Entities/Creature/Creature.cpp
#	src/server/game/Entities/Creature/Creature.h
#	src/server/game/Entities/GameObject/GameObject.h
#	src/server/game/Server/Packets/SpellPackets.cpp
#	src/server/game/Server/Packets/SpellPackets.h

Seyden committed:
Fix compile

Seyden committed:
Remove unused code

funjoker commented:
Error = TC-Error

funjoker commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2519,8 +2519,6 @@ float Creature::GetArmorMultiplierForTarget(WorldObject const* target) const
 
-    uint8 levelForTarget = GetLevelForTarget(target);
-
     return GetBaseArmorForLevel(unitTarget->getLevel()) / GetBaseArmorForLevel(getLevel());
```
return GetBaseArmorForLevel(unitTarget->getLevel()) / GetBaseArmorForLevel(getLevel());

Here not the armor of scaled level is returned?

Seyden replied:
Ofc not.. its a multiplier

joschiwald commented on src/server/game/Server/Packets/SpellPackets.cpp:
```diff
@@ -90,6 +90,76 @@ WorldPacket const* WorldPackets::Spells::SendUnlearnSpells::Write()
+    Expansion               = creatureTemplate->RequiredExpansion;
+    Class                   = creatureTemplate->unit_class;
+    TargetMinScalingLevel   = (uint8)creatureTemplate->levelScaling->minLevel;
+    TargetMaxScalingLevel   = (uint8)creatureTemplate->levelScaling->maxLevel;
```
no TargetScalingLevelDelta ?

funjoker replied:
Nope. Client scales visualy by itself.

Seyden committed:
UpperCamelCase fix

Seyden committed:
Fix fail in armor multiplier calculation

Seyden committed:
Attacker should never see the "original" damage value dealt to target.. moved it to DealDamage..

also should pets and other creatures that are in combat with the scaling creature, profit from it..

Seyden committed:
Build fix

Seyden committed:
Merge remote-tracking branch 'Trinity/master' into scaling

Seyden committed:
Merge remote-tracking branch 'Trinity/master' into scaling

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -1199,8 +1199,21 @@ void Creature::SelectLevel()
     // level
     uint8 minlevel = std::min(cInfo->maxlevel, cInfo->minlevel);
     uint8 maxlevel = std::max(cInfo->maxlevel, cInfo->minlevel);
```
move that into the if block too

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2450,17 +2470,87 @@ void Creature::AllLootRemovedFromCorpse()
+
+uint8 Creature::GetLevelForTarget(WorldObject const* target) const
+{
+    if (!target->ToUnit())
```
use your new IsUnit helper

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2450,17 +2470,87 @@ void Creature::AllLootRemovedFromCorpse()
+            return 1;
+        if (level > 255)
+            return 255;
+        return uint8(std::min(std::max(uint16(1), level), uint16(255)));
```
this is duplicate logic, use RoundToInterval

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2450,17 +2470,87 @@ void Creature::AllLootRemovedFromCorpse()
+        if (target->IsPlayer())
+            targetLevelWithDelta += target->GetUInt32Value(PLAYER_FIELD_SCALING_PLAYER_LEVEL_DELTA);
+
+        return std::min(std::max(targetLevelWithDelta, (uint8)GetUInt32Value(UNIT_FIELD_SCALING_LEVEL_MIN)), (uint8)GetUInt32Value(UNIT_FIELD_SCALING_LEVEL_MAX));
```
RoundToInterval

joschiwald commented on src/server/game/Server/Packets/SpellPackets.cpp:
```diff
@@ -90,6 +90,76 @@ WorldPacket const* WorldPackets::Spells::SendUnlearnSpells::Write()
+    TargetLevel             = target->getLevel();
+    Expansion               = creatureTemplate->RequiredExpansion;
+    Class                   = creatureTemplate->unit_class;
+    TargetMinScalingLevel   = (uint8)creatureTemplate->levelScaling->MinLevel;
```
why can TargetScaling depend on attacker while TargetLevel depend on target?

Seyden replied:
Because of the sniff.. dunno what blizz thought while they did this

joschiwald commented on src/server/game/Server/Packets/SpellPackets.cpp:
```diff
@@ -90,6 +90,76 @@ WorldPacket const* WorldPackets::Spells::SendUnlearnSpells::Write()
+{
+    // No scalable levels for any of them
+    if (   (!attacker->ToCreature() || !attacker->ToCreature()->HasScalableLevels())
+        && (!target->ToCreature()   || !target->ToCreature()->HasScalableLevels()))
```
this logic looks strange, isnt there a better way to do this?

Seyden replied:
Actually i don't know any better way.. if you have something in mind, let me know

joschiwald commented on sql/updates/world/master/9999_99_99_99_world.sql:
```diff
@@ -0,0 +1,9 @@
+  `LevelScalingMax` SMALLINT (5) UNSIGNED NOT NULL DEFAULT 0,
+  `LevelScalingDelta` SMALLINT (5) NOT NULL DEFAULT 0,
+  `VerifiedBuild` SMALLINT (5) DEFAULT NULL,
+  PRIMARY KEY (`entry`)
```
`Entry`

joschiwald commented on src/server/game/Entities/Object/Object.h:
```diff
@@ -255,26 +255,33 @@ class TC_GAME_API Object
+        inline bool IsAreaTrigger() const { return GetTypeId() == TYPEID_AREATRIGGER; }
+        AreaTrigger* ToAreaTrigger() { if (IsAreaTrigger()) return reinterpret_cast<AreaTrigger*>(this); else return NULL; }
+        AreaTrigger const* ToAreaTrigger() const { if (IsAreaTrigger()) return reinterpret_cast<AreaTrigger const*>(this); else return NULL; }
 
```
you forgot IsConversation

joschiwald commented on src/server/game/Entities/Object/Object.h:
```diff
@@ -255,26 +255,33 @@ class TC_GAME_API Object
-        Player const* ToPlayer() const { if (GetTypeId() == TYPEID_PLAYER) return reinterpret_cast<Player const*>(this); else return NULL; }
+        inline bool IsPlayer() const { return GetTypeId() == TYPEID_PLAYER; }
+        Player* ToPlayer() { if (IsPlayer()) return reinterpret_cast<Player*>(this); else return NULL; }
+        Player const* ToPlayer() const { if (IsPlayer()) return reinterpret_cast<Player const*>(this); else return NULL; }
```
while you are touching this lines, change NULL to nullptr

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -1199,8 +1199,21 @@ void Creature::SelectLevel()
+        SetLevel(cInfo->levelScaling->MaxLevel);
+        SetUInt32Value(UNIT_FIELD_SCALING_LEVEL_MIN, cInfo->levelScaling->MinLevel);
+        SetUInt32Value(UNIT_FIELD_SCALING_LEVEL_MAX, cInfo->levelScaling->MaxLevel);
+        SetUInt32Value(UNIT_FIELD_SCALING_LEVEL_DELTA, cInfo->levelScaling->DeltaLevel);
```
does this enable client side scaling?

Seyden replied:
Yes

joschiwald commented on src/server/game/Entities/Creature/Creature.cpp:
```diff
@@ -2450,17 +2470,87 @@ void Creature::AllLootRemovedFromCorpse()
+uint8 Creature::GetLevelForTarget(WorldObject const* target) const
+{
+    if (!target->ToUnit())
+        return Unit::GetLevelForTarget(target);
```
better, move that code at the end of the function and replace `return getLevel();`

Seyden committed:
Some changes

joschiwald committed:
some random changes

joschiwald committed:
calculate GetHealthMultiplierForTarget like the other multipliers

joschiwald committed:
datas is not a word

HannibalRoG commented:
And where would the information from the creature_template_scaling table come from? sniffer?

funjoker commented:
yes

HannibalRoG commented:
@seyden can you merge master branch into this PR?

joschiwald committed:
update packet

joschiwald committed:
prepared to merge

joschiwald committed:
small cleanup

joschiwald committed:
ordered includes

joschiwald committed:
fixed warning
